src/delayP.js

```diff
@@ -0,0 +1,43 @@
+/**
+ *
+ * @func delayP
+ * @memberOf RA
+ * @category Function
+ * @sig a -> Promise a
+ * @param {Number|Object} Argument to be resolved after a certain timeout and returns value if provided.
+ * @return {Promise} A Promise that is resolved with the given value if provided after certain delay.
+ * @see {@link RA.delayP|delayP}
+ * @example
+ *
+ * RA.delayP(200); //=> Returns Promise(undefined) after 200ms
+ * RA.delayP({ timeout: 1000, value: 'hello world' }); //=> Returns Promise('hello world') after 1000ms
+ * RA.delayP.reject(100); //=> rejects after 100 milliseconds with `undefined` value
+ * RA.delayP.reject({ timeout: 100, value: new Error('error') }); //=> rejects after 100 milliseconds with `Error('error')` value
+ */
+
+const makeDelay = (value, delayInMS, promiseType) => {
```

Comment:
Cool, nice job for a first run! Not take a look at this implementation. It's a psedo code (I never run it), but the code make use of the tools we have already available in ramda (R) + ramda-adjnuct (RA) and make your implementation a little bit more concise and readable. It's also settings some sane defaults when objects with no properties that we expect is provided.
```js
const makeDelay = R.curry((settleFnPicker, opts) => {
  let timeout = opts;
  let value = undefined;

  if (RA.isPlainObject(opts)) {
    timeout = R.propOr(0, 'timeout', opts);
    value = R.propOr(value, 'value', opts);
  }

  return new Promise((...args) => {
    const settleFn = settleFnPicker(args);
    setTimeout(R.partial(settleFn, [value]), timeout);
  });
});

const delayP = makeDelay(R.nth(0));
delayP.reject = makeDelay(R.nth(1));
```

Reply:
> ```js
> const delayP = makeDelay(R.nth(0));
> delayP.reject = makeDelay(R.nth(1));
> ```

Can you explain what this piece of code is doing.

Reply:
Also can you please review how was my implementation.

Reply:
> Also can you please review how was my implementation.

It was doing fine, until you got error from CI about repeating code. You eliminated the problem by increasing the complexity of the solution by introducing another helper called `delayPTemplate`. At that point the solution become unnecessarily complex. You switched the eliminating of the repetition for the complexity.

Reply:
> Can you explain what this piece of code is doing.

This code is using concept called currying, one the most powerfull tools of FP.
```js
const add = curry((a, b) => a + b);
add(1)(2); //=> 3

const add1 = add(1);
add1(2); //=> 3
```
Now if you understand this look at the code I provided again and try to understand it.

Reply:
Yup I got the concept of `currying` in FP. 
I had this doubt as you are passing `R.nth(0)` in  `makeDelay(R.nth(0));` you are not specifying any list or string as per the ramda docs `nth()` returns the nth element of given list or string.

Reply:
`R.nth` is of arity two, which means.
```js
const pickZeroIndexValue = nth(0);
pickZeroIndexValue([1, 2]); //=> 1;
```
Then in our case `pickZeroIndexValue` is called `settleFnPicker` and is passed into `makeDelay` as first argument. When `makeDelay` is executed, `settleFnPicker` is used to consume arguments of the promise lamda `...args` which is an array and will pick the first value from it, which is the `resolve` function.

Observe:
```js
const argsFn = (...args) => console.dir(args); 
argsFn(1, 2, 3); //=> outputs [1, 2, 3]

Reply:
Also can you explain this line `const settleFn = settleFnPicker(args);`
How `makeDelay()` knows about when to use `resolve()` and `reject()`.

Reply:
`makeDelay` doesn't know anything about that. We configure it it by passing `settleFnPicker` as an argument. And `settleFnPicker` knows if it should to pick `resolve` (nth(0)) or `reject` (nth(1)).
```js
const settleFn = settleFnPicker(args); // settleFn at this point is either resolve or reject function
```